python/mlx/nn/layers/linear.py

```diff
@@ -111,16 +111,31 @@ def __init__(
             )
 
     def _extra_repr(self) -> str:
+        out, in2, in1 = self.weight.shape
         return (
-            f"input1_dims={self.weight.shape[2]}, input2_dims={self.weight.shape[3]}, "
-            f"output_dims={self.weight.shape[1]}, bias={'bias' in self}"
+            f"input1_dims={in1}, input2_dims={in2}, output_dims={out}, "
+            f"bias={'bias' in self}"
         )
 
     def __call__(self, x1: mx.array, x2: mx.array) -> mx.array:
-        x1 = mx.expand_dims(x1, axis=(-2, -3))
-        x2 = mx.expand_dims(x2, axis=(-2))
-        y = mx.squeeze(x1 @ self.weight, -2).swapaxes(-1, -2)
-        y = mx.squeeze(x2 @ y, -2)
+        # Normalize shapes
+        out, in2, in1 = self.weight.shape
+        xshape = x1.shape[:-1]
+        x1 = x1.reshape(-1, in1)
+        x2 = x2.reshape(-1, in2)
+
+        # Perform the bilinear transformation
+        w = self.weight.reshape(in2 * out, in1)
+        y = x1 @ w.T
+        y = y.reshape(-1, out, in2).swapaxes(-2, -1)
+        y = x2[:, None, :] @ y
```

Comment:
Nit: you could avoid reshaping `x2` twice if you reshape it to `x2.reshape(-1, 1, in2)` above.